docs/realtime.md

```diff
@@ -0,0 +1,145 @@
+
+### optimization paths (future)
+
+- **bandwith** Limiting the number of events sent by allowing the client to specified it is only interested in events matching a selector *(files app only care about changes in the files of the current folder view)*  
+- **nb connection** Instead of 1 socket / tab, we can probably make 1 socket / browser using some hackish combination of SharedWorker / iframe.postMessage and a client-side demultiplexer.
+- **both** No need for realtime if the user is not using the tab (for most usecases), we could cut the realtime feed depending on [Page Visibility API](https://www.w3.org/TR/2011/WD-page-visibility-20110602/)
+
+
+## Go/Stack architecture
+
+
+- We assume all couchdb changes will originate from the stack
+- Events are generated at the stack level
+- We do **NOT** rely on couchdb `_changes?continuous` nor `_db_udpates`
+
+We create a realtime.Event interface, which we call in couchdb & vfs
+We accept websocket connection and bind them to a realtime.Dispatcher object.
+
+### Small cozy version
+
+It all happens in RAM, realtime.Event are immediately transmited to the dispatcher.
+
+
+### Big cozy version (ie. multiple stack instance)
+
+Redis pub/sub ?
+
+
+## Websocket API
+
+We start with a normal websocket handshake.
+
+Websockets include a protocol description in handshake, the protocol described below is hereby named `io.cozy.websocket` version `1`
+
+Changes to the websocket protocol should be given versions, support for older version should be maintained when reasonable.
+
+```http
+GET /websocket HTTP/1.1
+Host: mycozy.example.com
+Upgrade: websocket
+Connection: Upgrade
+Origin: http://calendar.mycozy.example.com
+Authorization: Bearer xxAppOrAuthTokenxx=
+Sec-WebSocket-Key: x3JrandomLkh9GBhXDw==
+Sec-WebSocket-Protocol: io.cozy.websocket
+Sec-WebSocket-Version: 1
+```
+
+Then messages are sent using json
+```
+client > {"mehod": "SUBSCRIBE",
+          payload: {"type":"io.cozy.files", "include_docs": true}
+client > {"mehod": "SUBSCRIBE",
+          payload: { "type":"io.cozy.contacts"}
+server > {"event": "change",
+          payload: {"id": "idA", "rev": "2-705...", type:"io.cozy.contacts"}}
+server > {"event": "change",
+          payload: {"id": "idA", "rev": "3-541...", deleted: true, type:"io.cozy.contacts"}}
+server > {"event": "change",
+          payload: {"id": "idB", "rev": "6-457...", type:"io.cozy.files", doc: {embeded doc ...}}}
+```
+
+### SUBSCRIBE
+
+A client can send a SUBSCRIBE request to be notified of changes.
+The payload is a selector for the events it wishes to receive
+For now the only possible selector is on type.
+
+`{"mehod": "SUBSCRIBE", payload: {type:"[desired doctype]"}`
+
+If the client wants to receive full documents with each events, it can include an `include_docs:true` parameter in the payload.
```

Comment:
How do we handle the permissions with the real-time? The payload for subscribe should match a permission rule?